libfwupdplugin/fu-progress.c

```diff
@@ -0,0 +1,714 @@
+	}
+	g_string_append(result, "-1 ]");
+	g_printerr("\n\n%s at %s\n\n", result->str, priv->id);
+	g_string_free(result, TRUE);
+}
+
+/**
+ * fu_progress_step_done:
+ * @self: A #FuProgress
+ *
+ * Called when the step_now sub-task has finished.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_progress_step_done(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	gdouble percentage;
+
+	g_return_if_fail(FU_IS_PROGRESS(self));
+
+	/* did we call done on a self that did not have a size set? */
+	if (priv->step_max == 0) {
+		g_autoptr(GString) str = g_string_new(NULL);
+		fu_progress_build_parent_chain(self, str, 0);
+		g_warning("progress done when no size set! [%s]: %s", priv->id, str->str);
+		return;
+	}
+
+	/* save the duration in the array */
+	if (priv->profile) {
+		gdouble elapsed = g_timer_elapsed(priv->timer, NULL);
+		if (priv->step_profile != NULL)
+			priv->step_profile[priv->step_now] = elapsed;
+		g_timer_start(priv->timer);
+	}
+
+	/* is already at 100%? */
+	if (priv->step_now >= priv->step_max) {
+		g_autoptr(GString) str = g_string_new(NULL);
+		fu_progress_build_parent_chain(self, str, 0);
+		g_warning("already at 100%% [%s]: %s", priv->id, str->str);
+		return;
+	}
+
+	/* is child not at 100%? */
+	if (priv->child != NULL) {
+		FuProgressPrivate *child_priv = GET_PRIVATE(priv->child);
+		if (child_priv->step_now != child_priv->step_max) {
+			g_autoptr(GString) str = g_string_new(NULL);
+			fu_progress_build_parent_chain(priv->child, str, 0);
+			g_warning("child is at %u/%u step_max and parent done [%s]: %s",
+				  child_priv->step_now,
+				  child_priv->step_max,
+				  priv->id,
+				  str->str);
+			/* do not abort, as we want to clean this up */
+		}
+	}
+
+	/* another */
+	priv->step_now++;
+
+	/* find new percentage */
+	if (priv->step_data == NULL) {
+		percentage = fu_progress_discrete_to_percent(priv->step_now, priv->step_max);
+	} else {
+		/* this is cumalative */
+		percentage = priv->step_data[priv->step_now - 1];
+	}
+	fu_progress_set_percentage(self, (guint)percentage);
+
+	/* show any profiling stats */
+	if (priv->profile && priv->step_now == priv->step_max && priv->step_profile != NULL)
+		fu_progress_show_profile(self);
+
+	/* reset child if it exists */
+	if (priv->child != NULL)
+		fu_progress_reset(priv->child);
+}
+
+/**
+ * fu_progress_sleep:
+ * @self: a #FuProgress
+ * @delay_secs: the delay in seconds
+ *
+ * Sleeps, setting the device progress from 0..100% as time continues.
+ * The value is given in whole seconds as it does not make sense to show the
+ * progressbar advancing so quickly for durations of less than one second.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_progress_sleep(FuProgress *self, guint delay_secs)
+{
+	gulong delay_us_pc = (delay_secs * G_USEC_PER_SEC) / 100;
+
+	g_return_if_fail(FU_IS_PROGRESS(self));
+	g_return_if_fail(delay_secs > 0);
+
+	fu_progress_set_percentage(self, 0);
+	for (guint i = 0; i < 100; i++) {
+		g_usleep(delay_us_pc);
```

Comment:
So is the idea that all plugins that have sleep statements will instead have a fu_progress_sleep?

Reply:
I think so. Then we know how the time maps to the percentage. I'm also pondering making it be milliseconds instead, as whole seconds doesn't seem useful enough for most devices.

Reply:
sounds g=fine